Jetty 9.4.x 2932 switchable classloader for session attribute values

janbartel opened the pull request:
This pr also merges the branch jetty-9.4.x-2919-marshalling-authenticated-infinispan-sessions. The changes from 2919 were necessary in order to make the classloader switchable when deserializing sessions, so it made sense to merge those changes into this branch to make a single pr.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/JDBCSessionDataStore.java:
```diff
@@ -764,10 +727,9 @@ protected void doInsert (String id, SessionData data)
                 ByteArrayOutputStream baos = new ByteArrayOutputStream();
                 ObjectOutputStream oos = new ObjectOutputStream(baos);
-                oos.writeObject(data.getAllAttributes());
+                SessionData.serializeAttributes(data, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
OK

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/JDBCSessionDataStore.java:
```diff
@@ -795,7 +757,7 @@ protected void doUpdate (String id, SessionData data)
                 ByteArrayOutputStream baos = new ByteArrayOutputStream();
                 ObjectOutputStream oos = new ObjectOutputStream(baos);
-                oos.writeObject(data.getAllAttributes());
+                SessionData.serializeAttributes(data, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
OK

sbordet commented on tests/test-sessions/test-file-sessions/src/test/java/org/eclipse/jetty/server/session/FileTestHelper.java:
```diff
@@ -177,14 +177,9 @@ public static void createFile (String id, String contextPath, String vhost,
-                    oos.writeUTF(name);
-                    oos.writeObject(attributes.get(name));
-                }
+                SessionData.serializeAttributes(tmp, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
The ObjectOutputStream is wrapping a FileOutputStream - it's inappropriate to close the OOS because it will close the FOS.

sbordet replied:
Ok, so `DataOutputStream out` can be included in the try-with-resources at the beginning? Then `ObjectOutputStream oos` can be left as is.

janbartel replied:
Yes, it is.

sbordet commented on tests/test-sessions/test-gcloud-sessions/src/test/java/org/eclipse/jetty/gcloud/session/GCloudSessionTestSupport.java:
```diff
@@ -130,15 +130,15 @@ public void createSession (String id, String contextPath, String vhost,
             ObjectOutputStream oos = new ObjectOutputStream(baos);
-            oos.writeObject(attributes);
-            oos.flush();
+            SessionData.serializeAttributes(tmp, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
ok

sbordet commented on tests/test-sessions/test-jdbc-sessions/src/test/java/org/eclipse/jetty/server/session/JdbcTestHelper.java:
```diff
@@ -303,15 +305,15 @@ public static void insertSession (String id, String contextPath, String vhost,
                 ObjectOutputStream oos = new ObjectOutputStream(baos);
-                Map<String,Object> emptyMap = Collections.emptyMap();
-                oos.writeObject(emptyMap);
+                SessionData.serializeAttributes(tmp, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
ok

sbordet commented on tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/MongoTestHelper.java:
```diff
@@ -226,13 +225,13 @@ public static void createUnreadableSession (String id, String contextPath, Strin
+            SessionData tmp = new SessionData (id, contextPath, vhost, created, accessed, lastAccessed, maxIdle, attributes);
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            ObjectOutputStream oos = new ObjectOutputStream(baos);
+            SessionData.serializeAttributes(tmp, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
ok

sbordet commented on tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/MongoTestHelper.java:
```diff
@@ -269,6 +268,51 @@ public static void createSession (String id, String contextPath, String vhost,
+            SessionData tmp = new SessionData (id, contextPath, vhost, created, accessed, lastAccessed, maxIdle, attributes);
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            ObjectOutputStream oos = new ObjectOutputStream(baos);
+            SessionData.serializeAttributes(tmp, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
ok

sbordet commented on jetty-infinispan/src/main/java/org/eclipse/jetty/session/infinispan/SessionDataMarshaller.java:
```diff
@@ -0,0 +1,119 @@
+        out.writeLong("maxInactiveMs", sdata.getMaxInactiveMs());
+
+        ByteArrayOutputStream bout = new ByteArrayOutputStream();
+        ObjectOutputStream oout = new ObjectOutputStream(bout);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
ok

sbordet commented on jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionDataStore.java:
```diff
@@ -532,29 +517,17 @@ public void doStore(String id, SessionData data, long lastSaveTime) throws Excep
+        
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ObjectOutputStream oos = new ObjectOutputStream(baos);
+        SessionData.serializeAttributes(data, oos);
```
`ObjectOutputStream` not closed, use try-with-resources.

janbartel replied:
ok

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionDataStore.java:
```diff
@@ -76,47 +83,84 @@ public void initialize (SessionContext context) throws Exception
+            throw exception.get();
+        
+        return reference.get();
+    }
```
Or use a `FuturePromise`, which can do both the result or the exception.

sbordet replied:
@gregw but it's not passed to another thread, so there is no "future" - it's just a method call with a return value and an exception.

gregw replied:
It if wrapped as a Runnable so it can't throw. We know it is run by the same thread but there is no guarantee of that. So there is a future.

sbordet replied:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

sbordet replied:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

sbordet replied:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

sbordet replied:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

sbordet replied:
Well, it is our own internal stuff. If it is run by the same thread there is no need for a Future.
Otherwise the method should return a `CompletableFuture`.
Like this it's too ugly.

janbartel replied:
Sorry guys, nobody has made a convincing argument to change this code in any way: this method is in the base class for all *SessionDataStores. It is implemented in a parsimonious way with 2 AtomicReferences, one to capture the result and one to capture the exception: any of the other suggestions here result in a hell of lot more object creation for absolutely no gain whatsoever.

sbordet replied:
@janbartel what I suggest will have no object creation at all!

gregw replied:
Well perhaps we can change `void ServletContext.run()` to `Object ServerContext.run() throws Exception`, but a) that is not really something for this PR; b) that will probably require new/changed catches elsewhere, which we need to review.    For now, this PR replaces several occurrences of this pattern with a single one, so even if it is ugly, it is better.

janbartel commented on tests/test-sessions/test-file-sessions/src/test/java/org/eclipse/jetty/server/session/FileTestHelper.java:
```diff
@@ -223,28 +218,18 @@ public static boolean checkSessionPersisted (SessionData data)
-                }
-            }
+            SessionData tmp = new SessionData(id, contextPath, vhost, created, accessed, lastAccessed, maxIdle);
+            ClassLoadingObjectInputStream ois =  new ClassLoadingObjectInputStream(di);
```
Closing the CLOIS and  it is wrapping another stream, which should not be closed at that point.

sbordet replied:
Same as above, you can initialize all these streams in the initial try-with-resources?
```java
try (FileInputStream in = new FileInputStream(file);
        DataInputStream di = new DataInputStream(in);
        ClassLoadingObjectInputStream ois =  new ClassLoadingObjectInputStream(di)) 
{
    ...
}
```

janbartel replied:
No, that cannot be done. When the CLOIS is created, it tries to read some bytes from the stream. It's not correct for those bytes to be read at that point, as it will screw up the deserialization.

janbartel commented on tests/test-sessions/test-mongodb-sessions/src/test/java/org/eclipse/jetty/nosql/mongodb/MongoTestHelper.java:
```diff
@@ -168,26 +170,23 @@ public static boolean checkSessionPersisted (SessionData data)
+            assertNotNull(attributes);
+            SessionData tmp = new SessionData(data.getId(), data.getContextPath(), data.getVhost(), created.longValue(), accessed.longValue(), lastAccessed.longValue(), maxInactive.longValue());
+            ByteArrayInputStream bais = new ByteArrayInputStream(attributes);
+            ClassLoadingObjectInputStream ois = new ClassLoadingObjectInputStream(bais);
```
See comment for FileTestHelper.

sbordet replied:
`ClassLoadingObjectInputStream ois` depends on the `ByteArrayInputStream bais` created the line before.
Both should be wrapped into try-with-resources.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/FileSessionDataStore.java:
```diff
@@ -685,8 +645,8 @@ protected SessionData load (InputStream is, String expectedId)
             // Attributes
-            restoreAttributes(di, di.readInt(), data);
-
+            ClassLoadingObjectInputStream ois =  new ClassLoadingObjectInputStream(is);
```
CLOIS wraps a stream that is passed into the method, not appropriate to close it here.

sbordet replied:
Ah, yes. Discard my comment.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/FileSessionDataStore.java:
```diff
@@ -277,80 +273,53 @@ public void sweepFile (long now, Path p)
+        }
+        catch (UnreadableSessionDataException e)
+        {
+            if (isDeleteUnrestorableFiles() && file.exists() && file.getParentFile().equals(_storeDir))
```
Not necessarily: I guess its possible that the scavenger thread has since deleted the file.

sbordet replied:
Then also `new FileInputStream(file)` can fail for the same reason, and there seem to be no logic to handle that.

The last expression `file.getParentFile().equals(_storeDir)` is always `true`, no?

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/FileSessionDataStore.java:
```diff
@@ -371,6 +340,7 @@ public void doStore(String id, SessionData data, long lastSaveTime) throws Excep
             }
             catch (Exception e)
             { 
+                e.printStackTrace();
```
oops.

janbartel commented on jetty-security/src/main/java/org/eclipse/jetty/security/authentication/FormAuthenticator.java:
```diff
@@ -195,6 +195,7 @@ public UserIdentity login(String username, Object password, ServletRequest reque
             HttpSession session = ((HttpServletRequest)request).getSession(true);
             Authentication cached=new SessionAuthentication(getAuthMethod(),user,password);
             session.setAttribute(SessionAuthentication.__J_AUTHENTICATED, cached);
+            session.setAttribute("SERVERCL", new MultiMap());
```
oops.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/AbstractSessionDataStore.java:
```diff
@@ -76,47 +83,84 @@ public void initialize (SessionContext context) throws Exception
+
+        _context.run(r);
+        if (exception.get() != null)
+            throw exception.get();
```
See my previous comment.

sbordet replied:
We have the chance to clean the API, so we don't want to use AtomicReferences to capture return values and exceptions when Java supports method calls!

janbartel commented on jetty-gcloud/jetty-gcloud-session-manager/src/main/java/org/eclipse/jetty/gcloud/session/GCloudSessionDataStore.java:
```diff
@@ -891,7 +871,7 @@ protected Entity entityFromSession (SessionData session, Key key) throws Excepti
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
         ObjectOutputStream oos = new ObjectOutputStream(baos);
-        oos.writeObject(session.getAllAttributes());
+        SessionData.serializeAttributes(session, oos);
```
ok

janbartel commented on jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/SessionDataSerializer.java:
```diff
@@ -0,0 +1,109 @@
+        out.writeLong(data.getMaxInactiveMs());
+        
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ObjectOutputStream oos = new ObjectOutputStream(baos);
```
ok

janbartel commented on jetty-memcached/jetty-memcached-sessions/src/main/java/org/eclipse/jetty/memcached/session/MemcachedSessionDataMap.java:
```diff
@@ -50,6 +53,66 @@
+            }
+            return rv;
+        }
+    }
```
ok

janbartel commented on jetty-nosql/src/main/java/org/eclipse/jetty/nosql/mongodb/MongoSessionDataStore.java:
```diff
@@ -178,107 +184,96 @@ public DBCollection getDBCollection ()
-                    exception.set(new UnreadableSessionDataException(id, _context, e));
+                    //attributes have special serialized format
+                    ByteArrayInputStream bais = new ByteArrayInputStream(attributes);
+                    ClassLoadingObjectInputStream ois = new ClassLoadingObjectInputStream(bais);
```
ok

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/session/FileSessionDataStore.java:
```diff
@@ -528,14 +492,8 @@ protected void save(OutputStream os, String id, SessionData data)  throws IOExce
-            oos.writeUTF(name);
-            oos.writeObject(data.getAttribute(name));
-        }
+        SessionData.serializeAttributes(data, oos);
```
OOS is wrapping a DataInputStream, which I don't want closed at this point. The outer DataInputStream will be closed.

sbordet replied:
Ah, yes. Discard my comment.

janbartel commented on jetty-memcached/jetty-memcached-sessions/pom.xml:
```diff
@@ -19,7 +19,7 @@
       <groupId>com.googlecode.xmemcached</groupId>
       <artifactId>xmemcached</artifactId>
-      <version>2.0.0</version>
+      <version>2.4.5</version>
```
Not sure what you mean here. How can anything in a pom affect anything in a module? Or do you mean to process the mod file and perform some kind of search and replacement with a maven plugin?

sbordet replied:
Ah! Good point, discard my comment.

janbartel commented:
Updated this PR to change the way we marshall the attributes of the session in infinispan:  because later versions of infinispan use an internal threadpool that knows nothing of our thread context classloader, the previous solution implemented in this PR wouldn't work when we upgraded infinispan.  The new solution always does our own serialization/deserialization of the attributes, so as far as infinispan is concerned they're just a byte array.  Also tested the session converter again and added a  "-Dverbose=true/false" option to print the actual stack traces obtained when conversion fails.

janbartel commented:
@sbordet  can you take another look at the change I just made? The reason for the change is that sessions that had been authenticated and stored in infinispan were not recoverable (which is the whole reason for introducing the new format and deserialization for infinispan). I made a tweak to the o.e.j.security.authentication.SessionAuthentication object that allows these legacy sessions to be deserialized for the purposes of conversation only.  This way, we should be able to convert all sessions to the new format that have been previously stored in infinispan, not just the un-authenticated ones.

blop commented:
I think this PR might break the HazelcastSessionDataStore : 

com.hazelcast.nio.serialization.HazelcastSerializationException: java.io.IOException: Not ClassLoadingObjectInputStream
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleException(SerializationUtil.java:63)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:193)
	at com.hazelcast.map.impl.proxy.MapProxySupport.toObject(MapProxySupport.java:1245)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.remove(MapProxyImpl.java:268)
	at org.eclipse.jetty.hazelcast.session.HazelcastSessionDataStore.delete(HazelcastSessionDataStore.java:77)
	at org.eclipse.jetty.server.session.AbstractSessionCache.loadSession(AbstractSessionCache.java:476)
	at org.eclipse.jetty.server.session.AbstractSessionCache.get(AbstractSessionCache.java:364)
	at org.eclipse.jetty.server.session.SessionHandler.getSession(SessionHandler.java:933)
	at org.eclipse.jetty.server.session.SessionHandler.getHttpSession(SessionHandler.java:538)
